### PR TITLE
Resolving Circular Imports in Ray Train

### PR DESCRIPTION
**Pull request by JasonLi1909**
### Resolving Circular Imports in Ray Train

**Problem**: There exists circular dependencies in Ray Train that cause circular import errors. `ray.train.v2` depends on `ray.train` for subclassing and overall base functionality while `ray.train` depends on `ray.train.v2` for patching of functionality when `RAY_TRAIN_V2_ENABLED=1`. This leads to import errors in the case when v2 attributes are imported directly which can occur by user misuse of APIs and during deserialization of v2 attributes during train worker setup. 

This PR resolves the issue with the following changes: 
- Importing ray train v1 modules within ray train v2 init files. This resolves circular import errors. 
- A comprehensive pre-commit linter that detects potential circular imports by parsing the v1 and v2 train directories, ensuring that for each v2 patch within the v1 `__init__.py` files, there are either (1) no circular imports in the overriding v2 files or (2) an import on the direct import path of the v2 file that suppresses the circular import
- A README.md for the linter describing the problem, the solution, and linter specification
- Tests for the linter 
- Tests for circular imports

Linter Notes: 

The linter takes a `patch_dir` to compare against the ray train directory for circular imports, allowing it to be easily extensible to other patching directories. It works by parsing the train directory and the patching directory for circular imports whether that is in the form of import-from, relative imports, or import statements. 

We choose to build a custom linter because while there are libraries like pycycle, there aren't any that can find dependencies between two separate directories which will be a requirement when extending this to other patching directories. Also, building a custom linter allows us to enforce the fix to resolve the issues. Overall, it allows us to tailor to our use case and is more lightweight.

Example Output via pre-commit:
Screenshot 2025-09-30 at 5 03 42 PM

---

> [!NOTE]
> Adds a circular-import linter for Ray Train, integrates it into pre-commit/CI, adds tests, and updates v2 package inits to suppress v1/v2 cycles.
> 
> - **Linting**:
>   - Add `python/ray/train/lint/check_circular_imports.py` to detect Train v1/v2 circular imports and reexports.
>   - Wire into pre-commit via local hook `check-train-circular-imports` in `.pre-commit-config.yaml` (args: `--patch_dir ray/train/v2`).
> - **Tests**:
>   - Add linter unit tests in `python/ray/train/v2/tests/test_circular_import_linter.py`.
>   - Add runtime import coverage in `python/ray/train/v2/tests/test_circular_imports.py`.
>   - Update `python/ray/train/v2/BUILD.bazel` to include `py_test(name="test_circular_imports")`.
> - **Package init updates (cycle suppression)**:
>   - Add `import ray.train.torch` in `python/ray/train/v2/torch/__init__.py`.
>   - Add `import ray.train.horovod` in `python/ray/train/v2/horovod/__init__.py`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2d538e5329840ea6ccc32f92087025ad01497ddd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: python/ray/train/v2/horovod/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+import ray.train.horovod
```
Comment (xinyuangui2) [inline python/ray/train/v2/horovod/__init__.py]: new line

**Review by xinyuangui2 — COMMENTED** [review]
Glad we resolved the circular import issue systematically.

File: python/ray/train/lint/check_circular_imports.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+            return module_str
+
+        package_parts = self._get_package_parts()
+
+        # If the relative import is out of bounds
+        if level - 1 > len(package_parts):
+            return None
+
+        # Base parts based on the level
+        base_module_parts = package_parts[: -((level - 1))]
```
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: good catch, fixing

File: python/ray/train/lint/check_circular_imports.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        patch_train_dir.rglob("__init__.py")
+    )
+    base_dir = get_base_dir()
+    for init_file in package_files:
+        relative_path = init_file.relative_to(base_dir)
+        dotted_module = str(relative_path.parent).replace("/", ".")
+        TRAIN_PACKAGES.add(dotted_module)
+
+
+def is_package(module_str: str) -> bool:
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py]: Is this used?

File: python/ray/train/lint/check_circular_imports.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    """
+
+    def __init__(self, module_name: str, is_package: bool) -> None:
+        self._module_name = module_name
+        self._is_package = is_package
+        self.imports: Set[Import] = set()
+
+    # --- private helpers ---
+
+    def _is_type_checking_test(self, expr: ast.AST) -> bool:
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py:81]: Does this still work if:
* Users use a variable that happens to be called TYPE_CHECKING
* Users import TYPE_CHECKING as something else?

File: python/ray/train/lint/check_circular_imports.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+        package_parts = self._get_package_parts()
+
+        # If the relative import is out of bounds
+        if level - 1 > len(package_parts):
+            return None
+
+        # Base parts based on the level
+        base_module_parts = (
+            package_parts if level == 1 else package_parts[: -((level - 1))]
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py]: nit: extra ()?

File: python/ray/train/lint/check_circular_imports.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+        # Skip class contents
+        return
+
+
+def collect_imports(
+    module_name: str, is_package: bool, source_text: str
+) -> Set[Import]:
+    try:
+        tree = ast.parse(source_text)
+    except SyntaxError:
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py:182]: Maybe print a warning here?

File: python/ray/train/lint/check_circular_imports.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+def get_base_dir() -> Path:
+    """Return the filesystem path to the ray python directory."""
+    try:
+        import ray
+
+        package_dir = Path(os.path.dirname(ray.__file__)).parent
+        return package_dir
+    except ModuleNotFoundError:
+        raise ModuleNotFoundError("ray package not found")
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py]: nit: why do we reraise here? What's the original error message?

File: python/ray/train/lint/check_circular_imports.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+                continue
+
+            # Skip if the patch train module init file imports the base train init module
+            patch_init_module = (
+                ".".join(patch_module.split(".")[:-1])
+                if not is_package(patch_module)
+                else patch_module
+            )
+            patch_init_imports = patch_train_init_imports.get(patch_init_module, set())
+            if any(base_train_init_module in imp.module for imp in patch_init_imports):
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py]: Are the containment checks here and elsewhere resistant to stuff like `ray.train` being part of `non_ray.train_foo`? If so would be worth using a more robust utility instead.

File: python/ray/train/lint/check_circular_imports.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+        self.is_package = is_package
+        self.module = module
+        self.names = names if names else []
+
+
+class ImportCollector(ast.NodeVisitor):
+    """
+    An AST node visitor that collects all module-level imports from a Python source file.
+    It traverses the AST and records module-level import statements (`import ...` and `from ... import ...`) that are not
+    inside function or class definitions, and that are not guarded by `if TYPE_CHECKING` or `if typing.TYPE_CHECKING`
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py:69]: `that are not inside function or class definitions`

Does this mean we can still get circular import errors from imports within function/class definitions?

File: python/ray/train/lint/check_circular_imports.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+
+        # Get the imports from the patch train files
+        patch_train_files = convert_to_file_paths(imports)
+        patch_train_file_imports = get_file_module_imports(
+            patch_train_files, module_match_string="ray.train"
+        )
+
+        for patch_module, imports in patch_train_file_imports.items():
+            # Skip if the base train init module is in the import path of the patch module
+            if base_train_init_module in patch_module:
```
Comment (TimothySeah) [inline python/ray/train/lint/check_circular_imports.py]: I'm confused - will this break if `ray/train/v2` is the `patch_dir`? My understanding is
* `base_train_patching_imports`: key = `ray.train`, value = `ray.train.v2` imports from `ray/train/__init__.py`
* `patch_train_init_imports`: key = `ray.train.v2`, value = `ray.train` imports from `ray/train/v2.__init__.py`
* `patch_train_file_imports`: key = the actual file that was imported like `ray.train.v2.api.result`. value = `ray.train` imports from that file

Is that correct? If so, would `if base_train_init_module in patch_module:` be `if "ray.train" in "ray.train.v2"` and therefore always pass the validation?

**Review by TimothySeah — COMMENTED** [review]
Good stuff, this is very thorough!

A few comments on the PR description:
* Don't link to internal notion page on OSS PR
* `which can occur by user misuse of APIs `: @justinvyu / @matthewdeng does this mean that users should never import anything from v2?
* Maybe add some testing notes e.g. temporarily introduce circular dependency and show precommit's output

Other questions/comments:
* Generally LGTM; will take a look at the unit tests on the next pass.
* I wonder if we should also get someone from devprod to review this / see if other teams like Ray Data want to use this.
* Consider adding a README or at least a comment at the top of `check_circular_imports.py` that explains what this does. The PR description is good but users who fail this check and want to learn more will most likely just look at the file.

File: python/ray/train/lint/check_circular_imports.py (hunk 3 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: used only once but changed to now using it in a few places

File: python/ray/train/lint/check_circular_imports.py (hunk 7 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: removed reraise

File: python/ray/train/lint/check_circular_imports.py (hunk 8 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: Good call. Created a utility `does_overlap` that uses string method `startswith` to cover this edge case

File: python/ray/train/lint/check_circular_imports.py (hunk 9 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:69]: No, only import statements on the module level are parsed. This way we avoid false positives as imports within functions/classes run later

File: python/ray/train/lint/check_circular_imports.py (hunk 10 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: So to clarify each dict: 
- `base_train_patching_imports`: dict[str, List[Import]] maps train v1 init modules like `ray.train` to their imports like `ray.train.v2`
- `patch_train_init_imports`: dict[str, List[Import]] maps all train v2 init modules like `ray.train.v2.tensorflow` to their imports like `ray.train.tensorflow` 
- `patch_train_file_imports`: dict[str, List[Import]] maps train v2 modules like `ray.train.v2.tensorflow.tensorflow_trainer` to their imports like `ray.train.tensorflow` 

So this does not break because `base_train_init_module` is not always `ray.train`. But you're correct on the expected behavior for `base_train_init_module="ray.train"`- we want this to be skipped because if the base_train_init_module is in the import path of the patch_module, then the init file will always run first and there will be no circular import error and hence, no violation. But in the case where base_train_init_module is, say, `ray.train.tensorflow` and the patch_module is `ray.train.v2.tensorflow`, then there could be a violation- and in this case we do not enter the if-statement.

File: python/ray/train/lint/check_circular_imports.py (hunk 4 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:81]: Nice catch, this is an edge case. The alternative would be to check by executing code but that can be convoluted. Instead, I added an additional check for if `typing.TYPE_CHECKING` is imported in the file and an additional test case. 

I think the remaining edge case where `typing.TYPE_CHECKING` is imported but the user overwrites it with a different value is unlikely enough to dismiss because it'd also have to happen at the module level and chances are that anything within some type of `TYPE_CHECKING` conditional would not cause an import error because it is used for only type checking.

File: python/ray/train/v2/tests/test_circular_import_linter.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+        from .submod import thing
+
+        if TYPE_CHECKING:
+            import foo
+
+        def f():
+            import json
+
+        class C:
+            from pkg import a
```
Comment (TimothySeah) [inline python/ray/train/v2/tests/test_circular_import_linter.py]: should you also assert that a is not in imports?

File: python/ray/train/v2/tests/test_circular_import_linter.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+        """
+    )
+
+    imports = cci.collect_imports(
+        module_name="pkg.module", is_package=False, source_text=source
+    )
+    imports = [imp.module for imp in imports]
+    assert "os" in imports
+    assert any(mod == "submod" or mod.endswith(".submod") for mod in imports)
```
Comment (TimothySeah) [inline python/ray/train/v2/tests/test_circular_import_linter.py]: I'm a bit confused by this assertion - what do we expect to see in imports? submod.thing?

File: python/ray/train/v2/tests/test_circular_import_linter.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    assert "foo" not in imports
+    assert "json" not in imports
+
+
+def test_import_collector_excludes_type_checking_without_from_import():
```
Comment (TimothySeah) [inline python/ray/train/v2/tests/test_circular_import_linter.py]: nit: both this and `test_import_collector_excludes_non_module_level_and_type_checking` assert foo not in imports. Maybe remove that check from `test_import_collector_excludes_non_module_level_and_type_checking` to keep each unit test targeted.

File: python/ray/train/v2/tests/test_circular_import_linter.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    v2_dir.mkdir(parents=True, exist_ok=True)
+    v1_pkg_dir.mkdir(parents=True, exist_ok=True)
+
+    # Base v1 package init: imports a v2 module
+    (v1_pkg_dir / "__init__.py").write_text(
+        "from ray.train.v2.tensorflow.tensorflow_trainer import tensorflow_trainer\n"
+    )
+
+    # v2 module that (incorrectly) imports back into v1 package
+    (v2_dir / "tensorflow_trainer.py").write_text(
```
Comment (TimothySeah) [inline python/ray/train/v2/tests/test_circular_import_linter.py:116]: nit: maybe also add a `(v1_subpkg_dir / "tensorflow_trainer.py")` that contains `"from ray.train.tensorflow import something\n"` and verify that this does not cause a violation. Basically to test https://github.com/ray-project/ray/pull/56921/files#diff-29a4df675700332da66bd599fd1b3dc0a9167a0a3a71c47c67c5032a49901241R360

File: python/ray/train/v2/tests/test_circular_import_linter.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    (v1_pkg_dir / "__init__.py").write_text(
+        "from ray.train.v2.tensorflow import tensorflow_trainer\n"
+    )
+
+    # v2 module that (incorrectly) imports back into v1 package
+    (v2_dir / "tensorflow_trainer.py").write_text(
+        "from ray.train.tensorflow import something\n"
+    )
+
+    # v2 package init WITHOUT importing v1 package AND reexporting the v2 module (should trigger violation)
```
Comment (TimothySeah) [inline python/ray/train/v2/tests/test_circular_import_linter.py]: Might also be worth including a unit test for `expand_to_include_reexports` since the logic is kinda confusing.

Will this unit test fail if `check_violations` does not call `expand_to_include_reexports`? If not it would be good to include such a test.

**Review by TimothySeah — APPROVED** [review]
Should be good from me aside from a few nits and clarifications

File: python/ray/train/v2/tests/test_circular_import_linter.py (hunk 11 above)
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_import_linter.py]: Added assertion

File: python/ray/train/v2/tests/test_circular_import_linter.py (hunk 12 above)
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_import_linter.py]: This tests for the linter being able to resolve relative imports. I will change this to be more specific: `assert "pkg.submod" in imports` because this is in the `pkg` module.

File: python/ray/train/v2/tests/test_circular_import_linter.py (hunk 13 above)
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_import_linter.py]: Good point, removing because this seems like a pretty redundant test overall.

File: python/ray/train/v2/tests/test_circular_import_linter.py (hunk 14 above)
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_import_linter.py:116]: Added a separate test `test_check_no_violation_on_overlapping_import_path` for this to avoid overloading this test case.

File: python/ray/train/v2/tests/test_circular_import_linter.py (hunk 15 above)
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_import_linter.py]: Good point, added separate test `test_expand_to_exclude_reexports` to test `expand_to_include_reexports`

File: python/ray/train/lint/check_circular_imports.py [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+
+The core challenge stems from the fact that `ray.train.v2` relies on `ray.train` for foundational functionality, while `ray.train` depends on `ray.train.v2` for enhanced features when the `RAY_TRAIN_V2_ENABLED` flag is set. This bidirectional dependency can lead to import errors, especially when v2 attributes are directly imported or during the deserialization of v2 attributes in train worker setups.
+
+To mitigate these issues, the linter implements several key strategies:
+1. **Import Resolution**: By enforcing the importing of `ray.train` v1 modules within `ray.train.v2` init files, the linter guides developers in resolving circular import errors.
+2. **Comprehensive Detection**: The linter parses both the v1 and v2 train directories to detect potential circular imports. It ensures that for each v2 patch within the v1 `__init__.py` files, there are no circular imports in the overriding v2 files, or there is an import on the direct import path of the v2 file that suppresses the circular import.
+3. **Extensibility**: The linter is designed to be extensible, allowing it to be adapted for other patching directories beyond Ray Train V2.
```
Comment (matthewdeng) [inline python/ray/train/lint/check_circular_imports.py]: This was confusing to me. Recommend articulating what are the exact problem scenarios & the solution for each instead. Also, extensibility is a good attribute to have and call out for the linter, but not exactly a "strategy" IMO.

File: python/ray/train/lint/check_circular_imports.py [hunk 17]
```diff
@@ -0,0 +1,423 @@
...
+
+### Example Violation Suppression:
+- **Typical Suppression**:
+  - `ray/train/a/init.py`: `from ray.train.v2.a.submodule import foo`
+  - `ray/train/v2/a/init.py`: `import ray.train.a` -> suppression import
+  - `ray/train/v2/a/sub_module.py`: `import ray.train.a`
+- **Early Suppression**:
+  - `ray/train/a/init.py`: `from ray.train.v2.a.submodule import foo`
+  - `ray/train/v2/init.py`: `import ray.train.a` -> suppression import
+  - `ray/train/v2/a/sub_module.py`: `import ray.train.a`
```
Comment (matthewdeng) [inline python/ray/train/lint/check_circular_imports.py]: Can you make this more clear what these examples are showing?

File: python/ray/train/lint/check_circular_imports.py [hunk 18]
```diff
@@ -0,0 +1,423 @@
...
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--patch_dir",
+        default="ray/train/v2",
+        help="Path to the directory containing patching contents",
+    )
+    args = parser.parse_args()
```
Comment (matthewdeng) [inline python/ray/train/lint/check_circular_imports.py:362]: The main usage for this is for "patch"ing imports, but is that a requirement that needs to be named here? In other words, could this be more generalizable to just detect circular imports between dir1 and dir2?

File: python/ray/train/lint/check_circular_imports.py [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+
+    # Compute relevant paths
+    base_dir = get_base_dir()
+    base_train_dir = get_base_train_dir()
+    patch_train_dir = base_dir / Path(args.patch_dir)
+
+    # Initialize train packages
+    initialize_train_packages(base_train_dir, patch_train_dir)
```
Comment (matthewdeng) [inline python/ray/train/lint/check_circular_imports.py]: Add more comments to explain what the purpose of each step here. E.g. not obvious why we need to initialize all the packages.

General criteria for this should be that I can read the docs at the top of the file and the `main` method and get a sense of correctness for this file, without having to read the implementation of each individual method.

File: python/ray/train/lint/check_circular_imports.py (hunk 18 above)
Comment (matthewdeng) [reply python/ray/train/lint/check_circular_imports.py:362]: If we want to constrain it to base/patch terminology, I think it'd be helpful to include the definition of these in the documentation at the top of the file.

File: python/ray/train/v2/tests/test_circular_imports.py [hunk 20]
```diff
@@ -0,0 +1,53 @@
+def test_train_import():
```
Comment (matthewdeng) [inline python/ray/train/v2/tests/test_circular_imports.py:8]: Make sure to include pytest `__main__` logic and add to BUILD file to run these tests.

File: python/ray/train/v2/tests/test_circular_imports.py [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+    from ray.train.v2.api.callback import UserCallback  # noqa: F401
```
Comment (matthewdeng) [inline python/ray/train/v2/tests/test_circular_imports.py:9]: Document what this test file should be testing? Currently not obvious since there are no assertions.

File: python/ray/train/v2/tests/test_circular_imports.py (hunk 20 above)
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_imports.py:8]: Added test to BUILD file and included pytest `__main__` logic for both test files.

File: python/ray/train/lint/check_circular_imports.py (hunk 16 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: addressed in the new README.md

File: python/ray/train/lint/check_circular_imports.py (hunk 17 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: Removing these and addressed more clearly in the new README.md

File: python/ray/train/lint/check_circular_imports.py (hunk 18 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:362]: Some of the behavior of the linter might not necessarily be generalizable to other sub-directories. So will clarify the base/patch terminology in the new README.md

File: python/ray/train/lint/check_circular_imports.py (hunk 19 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: Done

File: python/ray/train/lint/check_circular_imports.py [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+
+    # Collect all packages under both base and patch train dirs
+    package_files = list(base_train_dir.rglob("__init__.py")) + list(
+        patch_train_dir.rglob("__init__.py")
+    )
+    base_dir = get_base_dir()
+    for init_file in package_files:
+        relative_path = init_file.relative_to(base_dir)
+        dotted_module = str(relative_path.parent).replace("/", ".")
+        TRAIN_PACKAGES.add(dotted_module)
```
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:26]: this allows for generalization to other patch directories

File: python/ray/train/lint/check_circular_imports.py [hunk 23]
```diff
@@ -0,0 +1,395 @@
...
+
+
+def get_base_train_dir() -> Path:
+    """Return the filesystem path to the ray train directory."""
+    return get_base_dir() / "ray/train"
+
+
+def does_overlap(main_module: str, module: str) -> bool:
+    """Checks if the init file of module is on the import path of main_module"""
+    return main_module.startswith(f"{module}.") or main_module == module
```
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:47]: not true

File: python/ray/train/v2/tests/test_circular_imports.py (hunk 21 above)
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_imports.py:9]: Done, left a description at the top of the file

File: python/ray/train/lint/check_circular_imports.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:47]: not true

File: python/ray/train/v2/tests/test_circular_import_linter.py [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+
+    imports = cci.collect_imports(
+        module_name="pkg.module", is_package=False, source_text=source
+    )
+    imports = [imp.module for imp in imports]
+    assert "os" in imports
+    assert "pkg.submod" in imports
+    assert "foo" not in imports
+    assert "json" not in imports
+    assert "pkg" not in imports
```
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/v2/tests/test_circular_import_linter.py:35]: updated test to only `import pkg`

File: python/ray/train/lint/check_circular_imports.py [hunk 25]
```diff
@@ -0,0 +1,395 @@
...
+        # Filter patch init file imports to those that only contain the right names
+        for base_import in base_imports:
+            if base_import.module in reexports:
+                import_list = reexports[base_import.module]
+                target_reexports = [
+                    imp
+                    for imp in import_list
+                    if set(imp.names) & set(base_import.names)
+                ]
+                agg_reexports.extend(target_reexports)
```
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:288]: Not possible, self.names is set to empty list upon init

File: python/ray/train/lint/check_circular_imports.py [hunk 26]
```diff
@@ -0,0 +1,395 @@
...
+            # Skip if the base train init module is in the import path of the patch module
+            if does_overlap(patch_module, base_train_init_module):
+                continue
+
+            # Skip if the patch train module init file imports the base train init module
+            patch_init_module = (
+                ".".join(patch_module.split(".")[:-1])
+                if not is_train_package(patch_module)
+                else patch_module
+            )
```
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:337]: not true

File: python/ray/train/lint/check_circular_imports.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:288]: it can't be None, it's always set to an empty list

File: python/ray/train/v2/torch/__init__.py [hunk 27]
```diff
@@ -0,0 +1,2 @@
+# Note: This import is to avoid circular import errors
+import ray.train.torch
```
Comment (matthewdeng) [inline python/ray/train/v2/torch/__init__.py]: Live by the lint, die by the lint.

```suggestion
import ray.train.torch  # noqa: F401
```

File: .pre-commit-config.yaml [hunk 28]
```diff
@@ -227,3 +227,14 @@ repos:
         language: python
         files: '^src/ray/'
         types: [c++]
+
+  - repo: local
+    hooks:
+    - id: check-train-circular-imports
+      name: Check ray train circular imports
```
Comment (matthewdeng) [inline .pre-commit-config.yaml]: nit: Capitalize library names everywhere so it's easier to read it as a proper noun.
```suggestion
      name: Check Ray Train circular imports
```

File: python/ray/train/lint/check_circular_imports.py [hunk 29]
```diff
@@ -0,0 +1,395 @@
...
+import ast
+import os
+import sys
+from pathlib import Path
+from typing import Dict, List, Optional, Set, Tuple
+
+TRAIN_PACKAGES = set()
+
+
+def initialize_train_packages(base_train_dir: Path, patch_train_dir: Path) -> None:
```
Comment (matthewdeng) [inline python/ray/train/lint/check_circular_imports.py]: I'd rename this to something like `find` or `detect`, because `initialize` is overused in the context of this PR 😄

File: python/ray/train/lint/README.md [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+
+When there is a circular dependency present in the codebase, import errors are triggered by importing directly from the patching directory. For example, directly importing the v2 TensorflowTrainer with `from ray.train.v2.tensorflow.tensorflow_trainer import TensorflowTrainer` rather than relying on the conventional routing logic via `from ray.train.tensorflow import TensorflowTrainer` results in a `ImportError: cannot import name TensorflowTrainer from partially initialized ray.train.v2.tensorflow.tensorflow_trainer`. This happens in the case of user API misuse or during the deserialization of the train function on train worker setup. The following image depicts the error path of such erroneous imports.
+
+![ErrorPath](./images/ErrorPath.png)
+
+### The Fix:
+
+To make Ray Train resilient to such erroneous imports, this linter proactively detects circular imports and specifies how to fix it. The fix perscribed by the linter prevents import errors by importing the base train packages early within in the patching directory. In the below example, the previously depicted circular import is resolved by the linter's suggested fix to import `ray.train.foo` early in `ray.train.v2.foo`.
```
Comment (matthewdeng) [inline python/ray/train/lint/README.md]: Can you call out why importing packages earlier fixes this problem? 

My understanding is that it's because you can import from a partially loaded `__init__.py` file/package, but not from the other modules, but I'm not sure the right way to articulate it...

**Review by matthewdeng — APPROVED** [review]

File: python/ray/train/lint/check_circular_imports.py (hunk 29 above)
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: good call, renaming to `find_train_packages`

File: python/ray/train/lint/README.md (hunk 30 above)
Comment (JasonLi1909) [reply python/ray/train/lint/README.md]: Added an additional sentence to explain the fix.

File: python/ray/train/lint/check_circular_imports.py [hunk 31]
```diff
@@ -0,0 +1,395 @@
...
+    # Collect all packages under both base and patch train dirs
+    package_files = list(base_train_dir.rglob("__init__.py")) + list(
+        patch_train_dir.rglob("__init__.py")
+    )
+    base_dir = get_base_dir()
+    for init_file in package_files:
+        relative_path = init_file.relative_to(base_dir)
+        dotted_module = str(relative_path.parent).replace("/", ".")
+        TRAIN_PACKAGES.add(dotted_module)
+
```
[the remark this answers is not in the document]
Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py:27]: good call, fixed

File: python/ray/train/lint/check_circular_imports.py [hunk 32]
```diff
@@ -0,0 +1,395 @@
...
+
+def is_train_package(module_str: str) -> bool:
+    return module_str in TRAIN_PACKAGES
+
+
+def get_base_dir() -> Path:
+    """Return the filesystem path to the ray python directory."""
+    import ray
```
Comment (matthewdeng) [inline python/ray/train/lint/check_circular_imports.py]: This assumes that `ray` is installed from source, which may not always be true. If you just install it via `pip install ray` it won't point to the local package.

Comment (JasonLi1909) [reply python/ray/train/lint/check_circular_imports.py]: Changed function to compute the base directory relative to the location of the `check_circular_imports.py` instead, no longer requiring `import ray`
